fdbserver/RESTKmsConnector.actor.cpp

```diff
@@ -289,25 +293,27 @@ ACTOR Future<Void> discoverKmsUrls(Reference<RESTKmsConnectorCtx> ctx, RefreshPe
 	return Void();
 }
 
-void checkResponseForError(Reference<RESTKmsConnectorCtx> ctx, const rapidjson::Document& doc, bool isCipher) {
+void checkResponseForError(Reference<RESTKmsConnectorCtx> ctx,
+                           const rapidjson::Document& doc,
+                           IsCipherType isCipherType) {
 	// check version tag sanity
 	if (!doc.HasMember(REQUEST_VERSION_TAG) || !doc[REQUEST_VERSION_TAG].IsInt()) {
-		TraceEvent(SevWarn, "KMSResponseMissingVersion", ctx->uid).log();
-		throw operation_failed();
-	} else {
-		const int version = doc[REQUEST_VERSION_TAG].GetInt();
-		const int maxSupportedVersion = isCipher ? SERVER_KNOBS->REST_KMS_CURRENT_CIPHER_REQUEST_VERSION
-		                                         : SERVER_KNOBS->REST_KMS_CURRENT_BLOB_METADATA_REQUEST_VERSION;
-		if (version == INVALID_REQUEST_VERSION || version > maxSupportedVersion) {
-			TraceEvent(SevWarn, "KMSResponseInvalidVersion", ctx->uid)
-			    .detail("Version", version)
-			    .detail("MaxSupportedVersion", maxSupportedVersion);
-			throw operation_failed();
-		}
+		TraceEvent(SevWarnAlways, "RESTKMSResponseMissingVersion", ctx->uid).log();
+		throw rest_malformed_response();
+	}
+
+	const int version = doc[REQUEST_VERSION_TAG].GetInt();
+	const int maxSupportedVersion = isCipherType ? SERVER_KNOBS->REST_KMS_CURRENT_CIPHER_REQUEST_VERSION
+	                                             : SERVER_KNOBS->REST_KMS_CURRENT_BLOB_METADATA_REQUEST_VERSION;
+	if (version == INVALID_REQUEST_VERSION || version > maxSupportedVersion) {
```

Comment:
Shall we define `MAX_SUPPORTED_VERSION` as a constant? Say if one day we support two versions, but setting the knob to v1, the check will reject a v2 response, which is in fact supported.

Reply:
Idea is the following, as with releases, MAX_SUPPORTED_VERSION can only be _incremented_ with time, for instance: our request/response payload version is `1`, given it is a json format, one can evolve the scheme by adding/removing field, if they do, they need to increment the KNOB value to reflect the fact (also update the code to handle it, off course). The json fields also allows the external KMS endpoint to evolve with time. 

However it is an interesting thought, for downward compatibility, KMS shouldn't reply with any future version i.e. addressing the usecase you mentioned system was running with V2 and then downgraded to V1; it should be done with care, one possible option is to force _cluster-recovery_, however, it still doesn't help with delayed TCP responses or re-transmits. 

One possible option is to maintain two knobs:
1. Current request version
2. Max supported request version

Invariant here would be `current <= max` for all cases (upgrade and/or downgrade) 

Sounds good?

Reply:
an additional knob to specify max supported version would work. Just making it a constant is simpler. Both sound good to me.

Reply:
I can make it constant, but, then it wouldn't solve if the KNOB values is updated dynamically causing random failures. Idea with two knobs to allow the following:
User can control the `current_supported_version` irrespective of `max_supported_version` without breaking upgradability and/or downgradability.